Mathlib/Algebra/MonoidAlgebra/MapDomain.lean

```diff
@@ -88,6 +92,104 @@ lemma mapRangeRingHom_comp_mapDomainRingHom (f : R →+* S) (g : M →* N) :
 
 end MonoidAlgebra
 
+namespace Equiv
+variable [Semiring R] [Mul M] [Mul N] [Mul O]
+
+/-- Equivalent monoids have additively isomorphic monoid algebras.
+
+`MonoidAlgebra.mapDomain` as an `AddEquiv`. -/
+@[to_additive (dont_translate := R) (attr := simps apply symm_apply)
+/-- Equivalent additive monoids have additively isomorphic additive monoid algebras.
+
+`AddMonoidAlgebra.mapDomain` as an `AddEquiv`. -/]
+def monoidAlgebraCongrRight (e : M ≃ N) : MonoidAlgebra R M ≃+ MonoidAlgebra R N where
+  toFun x := x.mapDomain e
+  invFun x := x.mapDomain e.symm
+  left_inv x := by ext; simp
+  right_inv x := by ext; simp
+  map_add' x y := by ext; simp
+
+@[simp] lemma symm_monoidAlgebraCongrRight (e : M ≃ N) :
+    e.monoidAlgebraCongrRight.symm = e.symm.monoidAlgebraCongrRight (R := R) := rfl
+
+@[simp] lemma monoidAlgebraCongrRight_trans (e₁ : M ≃ N) (e₂ : N ≃ O) :
+    (e₁.trans e₂).monoidAlgebraCongrRight (R := R) =
+      e₁.monoidAlgebraCongrRight.trans e₂.monoidAlgebraCongrRight := by
+  ext; simp [Finsupp.mapDomain_comp]
+
+end Equiv
+
+namespace AddEquiv
+variable [Semiring R] [Semiring S] [Semiring T] [Mul M]
+
+/-- Additively isomorphic rings have additively isomorphic monoid algebras.
+
+`Finsupp.mapRange` as an `AddEquiv`. -/
+@[to_additive (dont_translate := R S) (attr := simps)
+/-- Additively isomorphic rings have additively isomorphic additive monoid algebras.
+
+`Finsupp.mapRange` as an `AddEquiv`. -/]
+def monoidAlgebraCongrLeft (e : R ≃+ S) : MonoidAlgebra R M ≃+ MonoidAlgebra S M where
+  toFun x := .mapRange e e.map_zero x
+  invFun x := .mapRange e.symm e.symm.map_zero x
+  left_inv x := by ext; simp
+  right_inv x := by ext; simp
+  map_add' x y := by ext; simp
+
+@[simp] lemma symm_monoidAlgebraCongrLeft (e : R ≃+ S) :
+    e.monoidAlgebraCongrLeft.symm = e.symm.monoidAlgebraCongrLeft (M := M) := rfl
+
+@[simp] lemma monoidAlgebraCongrLeft_trans (e₁ : R ≃+ S) (e₂ : S ≃+ T) :
+    (e₁.trans e₂).monoidAlgebraCongrLeft (M := M) =
+      e₁.monoidAlgebraCongrLeft.trans e₂.monoidAlgebraCongrLeft := by ext; simp
+
+end AddEquiv
+
+namespace MulEquiv
+variable [Semiring R] [Monoid M] [Monoid N] [Monoid O]
+
+/-- Isomorphic monoids have isomorphic monoid algebras. -/
+@[to_additive (dont_translate := R) (attr := simps! apply symm_apply)
+/-- Isomorphic monoids have isomorphic additive monoid algebras. -/]
+def monoidAlgebraCongrRight (e : M ≃* N) : MonoidAlgebra R M ≃+* MonoidAlgebra R N :=
```

Comment:
This is a weaker version of `MonoidAlgebra.domCongr`; do we need both?

Reply:
`MonoidAlgebra.domCongr` is also defined too late to be used in the places where I want to use without increasing imports

Reply:
Not without removing some `assert_not_exists`s. But in other places of the library we tend to keep the `RingEquiv` versions even when they are not more general than the `AlgEquiv` ones, so I claim this is okay.